src/pymor/algorithms/line_search.py

```diff
@@ -56,15 +60,22 @@ def armijo(f, starting_point, direction, grad=None, initial_value=None,
     slope = 0.0
 
     # Compute slope if gradient is provided
-    if grad:
-        slope = min(grad.inner(direction), 0.0)
+    if grad is not None:
+        slope = min(np.inner(grad, direction), 0.0)
+
+    if additional_criterion is not None:
+        assert callable(additional_criterion)
+    else:
+        additional_criterion = lambda *args: False
 
     while True:
         # Compute new function value
         current_value = f(starting_point + alpha * direction)
         # Check the Armijo condition
-        if current_value < initial_value + alpha * beta * slope:
+        if (current_value < initial_value + alpha * beta * slope
+                or additional_criterion(f, starting_point, alpha, direction, grad, slope)):
```

Comment:
```suggestion
        if (current_value < initial_value + alpha * beta * slope
                or additional_criterion(current_value, new_mu)):
```
and `new_mu` defined as in line 73